File: lib/features/notifications/domain/notification.dart
enum NotificationChannel { inApp, email, push, sms, whatsapp }

enum NotificationFrequency { immediate, dailyDigest, weeklyDigest, disabled }

enum NotificationEventType {
  matchingOpportunity,
  deadlineReminder,
  requirementsChanged,
  deadlineChanged,
  opportunityVerified,
  savedOpportunityExpired,
  applicationProgress,
  providerAnnouncement,
  emergencySystemMessage,
}

enum NotificationDeliveryStatus {
  scheduled,
  queued,
  processing,
  sent,
  delivered,
  read,
  failed,
  retrying,
  cancelled,
  expired,
}

class NotificationPreferences {
  const NotificationPreferences({
    this.channels = const {
      NotificationChannel.inApp,
      NotificationChannel.email,
      NotificationChannel.push,
    },
    this.frequency = NotificationFrequency.immediate,
    this.reminderDays = const {30, 14, 7, 3, 1},
    this.matchingOpportunities = true,
    this.opportunityChanges = true,
    this.verificationUpdates = true,
    this.savedOpportunityExpiry = true,
    this.quietHoursStart,
    this.quietHoursEnd,
    this.timezone = 'UTC',
    this.dailyLimit = 10,
    this.groupNotifications = true,
    this.unsubscribedTypes = const {},
  });

  final Set<NotificationChannel> channels;
  final NotificationFrequency frequency;
  final Set<int> reminderDays;
  final bool matchingOpportunities;
  final bool opportunityChanges;
  final bool verificationUpdates;
  final bool savedOpportunityExpiry;
  final int? quietHoursStart;
  final int? quietHoursEnd;
  final String timezone;
  final int dailyLimit;
  final bool groupNotifications;
  final Set<NotificationEventType> unsubscribedTypes;

  NotificationPreferences copyWith({
    Set<NotificationChannel>? channels,
    NotificationFrequency? frequency,
    Set<int>? reminderDays,
    bool? matchingOpportunities,
    bool? opportunityChanges,
    bool? verificationUpdates,
    bool? savedOpportunityExpiry,
    int? quietHoursStart,
    int? quietHoursEnd,
    String? timezone,
    int? dailyLimit,
    bool? groupNotifications,
    Set<NotificationEventType>? unsubscribedTypes,
  }) => NotificationPreferences(
    channels: channels ?? this.channels,
    frequency: frequency ?? this.frequency,
    reminderDays: reminderDays ?? this.reminderDays,
    matchingOpportunities: matchingOpportunities ?? this.matchingOpportunities,
    opportunityChanges: opportunityChanges ?? this.opportunityChanges,
    verificationUpdates: verificationUpdates ?? this.verificationUpdates,
    savedOpportunityExpiry:
        savedOpportunityExpiry ?? this.savedOpportunityExpiry,
    quietHoursStart: quietHoursStart ?? this.quietHoursStart,
    quietHoursEnd: quietHoursEnd ?? this.quietHoursEnd,
    timezone: timezone ?? this.timezone,
    dailyLimit: dailyLimit ?? this.dailyLimit,
    groupNotifications: groupNotifications ?? this.groupNotifications,
    unsubscribedTypes: unsubscribedTypes ?? this.unsubscribedTypes,
  );
}

class NotificationTemplate {
  const NotificationTemplate({
    required this.id,
    required this.type,
    required this.titleTemplate,
    required this.bodyTemplate,
    required this.channels,
  });
  final String id;
  final NotificationEventType type;
  final String titleTemplate;
  final String bodyTemplate;
  final Set<NotificationChannel> channels;
}

class ScholarSphereNotification {
  const ScholarSphereNotification({
    required this.id,
    required this.userId,
    required this.type,
    required this.title,
    required this.message,
    required this.channels,
    required this.scheduledFor,
    required this.createdAt,
    this.opportunityId,
    this.readAt,
    this.templateId,
    this.relatedEntityType,
    this.relatedEntityId,
    this.sentAt,
    this.deliveredAt,
    this.status = NotificationDeliveryStatus.scheduled,
    this.retryCount = 0,
    this.failureReason,
    this.timezone = 'UTC',
    this.groupKey,
  });

  final String id;
  final String userId;
  final NotificationEventType type;
  final String title;
  final String message;
  final Set<NotificationChannel> channels;
  final DateTime scheduledFor;
  final DateTime createdAt;
  final String? opportunityId;
  final DateTime? readAt;
  final String? templateId;
  final String? relatedEntityType;
  final String? relatedEntityId;
  final DateTime? sentAt;
  final DateTime? deliveredAt;
  final NotificationDeliveryStatus status;
  final int retryCount;
  final String? failureReason;
  final String timezone;
  final String? groupKey;

  bool get isRead => readAt != null;

  ScholarSphereNotification markRead(DateTime at) => ScholarSphereNotification(
    id: id,
    userId: userId,
    type: type,
    title: title,
    message: message,
    channels: channels,
    scheduledFor: scheduledFor,
    createdAt: createdAt,
    opportunityId: opportunityId,
    readAt: at,
    templateId: templateId,
    relatedEntityType: relatedEntityType,
    relatedEntityId: relatedEntityId,
    sentAt: sentAt,
    deliveredAt: deliveredAt,
    status: NotificationDeliveryStatus.read,
    retryCount: retryCount,
    failureReason: failureReason,
    timezone: timezone,
    groupKey: groupKey,
  );

  ScholarSphereNotification copyWith({
    NotificationDeliveryStatus? status,
    DateTime? sentAt,
    DateTime? deliveredAt,
    int? retryCount,
    String? failureReason,
    DateTime? scheduledFor,
  }) => ScholarSphereNotification(
    id: id,
    userId: userId,
    type: type,
    title: title,
    message: message,
    channels: channels,
    scheduledFor: scheduledFor ?? this.scheduledFor,
    createdAt: createdAt,
    opportunityId: opportunityId,
    readAt: readAt,
    templateId: templateId,
    relatedEntityType: relatedEntityType,
    relatedEntityId: relatedEntityId,
    sentAt: sentAt ?? this.sentAt,
    deliveredAt: deliveredAt ?? this.deliveredAt,
    status: status ?? this.status,
    retryCount: retryCount ?? this.retryCount,
    failureReason: failureReason ?? this.failureReason,
    timezone: timezone,
    groupKey: groupKey,
  );
}

class NotificationDeliveryAnalytics {
  const NotificationDeliveryAnalytics({
    required this.total,
    required this.delivered,
    required this.read,
    required this.failed,
    required this.retried,
  });
  final int total;
  final int delivered;
  final int read;
  final int failed;
  final int retried;
}

File: scholarsphere_backend/app/models/notification.py
import enum
from datetime import datetime
from uuid import UUID

from sqlalchemy import Boolean, DateTime, Enum, ForeignKey, Index, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base
from app.models.external_opportunity import JSONType


class NotificationChannel(str, enum.Enum):
    in_app = "in_app"
    email = "email"
    push = "push"
    sms = "sms"
    whatsapp = "whatsapp"


class NotificationFrequency(str, enum.Enum):
    immediate = "immediate"
    daily_digest = "daily_digest"
    weekly_digest = "weekly_digest"
    disabled = "disabled"


class NotificationEventType(str, enum.Enum):
    matching_opportunity = "matching_opportunity"
    deadline_reminder = "deadline_reminder"
    requirements_changed = "requirements_changed"
    deadline_changed = "deadline_changed"
    opportunity_verified = "opportunity_verified"
    saved_opportunity_expired = "saved_opportunity_expired"
    application_progress = "application_progress"
    provider_announcement = "provider_announcement"
    emergency_system_message = "emergency_system_message"


class NotificationDeliveryStatus(str, enum.Enum):
    scheduled = "scheduled"
    queued = "queued"
    processing = "processing"
    sent = "sent"
    delivered = "delivered"
    read = "read"
    failed = "failed"
    retrying = "retrying"
    cancelled = "cancelled"
    expired = "expired"


class NotificationPreferences(Base):
    """One row per user - a settings table, not an event table, so the

    Firebase uid is the primary key itself rather than a separate uuid.
    """

    __tablename__ = "notification_preferences"

    user_id: Mapped[str] = mapped_column(String(255), primary_key=True)
    channels: Mapped[list[str]] = mapped_column(
        JSONType,
        default=lambda: [
            NotificationChannel.in_app.value,
            NotificationChannel.email.value,
            NotificationChannel.push.value,
        ],
        nullable=False,
    )
    frequency: Mapped[NotificationFrequency] = mapped_column(
        Enum(NotificationFrequency, native_enum=False),
        default=NotificationFrequency.immediate,
        nullable=False,
    )
    reminder_days: Mapped[list[int]] = mapped_column(
        JSONType, default=lambda: [30, 14, 7, 3, 1], nullable=False
    )
    matching_opportunities: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    opportunity_changes: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    verification_updates: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    saved_opportunity_expiry: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    quiet_hours_start: Mapped[int | None] = mapped_column(Integer)
    quiet_hours_end: Mapped[int | None] = mapped_column(Integer)
    timezone: Mapped[str] = mapped_column(String(64), default="UTC", nullable=False)
    daily_limit: Mapped[int] = mapped_column(Integer, default=10, nullable=False)
    group_notifications: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    unsubscribed_types: Mapped[list[str]] = mapped_column(JSONType, default=list, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )


class NotificationTemplate(Base):
    """Admin-managed catalog, upserted by a client-chosen id (not server-generated)."""

    __tablename__ = "notification_templates"

    id: Mapped[str] = mapped_column(String(255), primary_key=True)
    type: Mapped[NotificationEventType] = mapped_column(
        Enum(NotificationEventType, native_enum=False), nullable=False
    )
    title_template: Mapped[str] = mapped_column(String(500), nullable=False)
    body_template: Mapped[str] = mapped_column(Text, nullable=False)
    channels: Mapped[list[str]] = mapped_column(JSONType, default=list, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )


class ScholarSphereNotification(Base):
    """A scheduled/sent notification instance.

    ``id`` is a plain string, not a UUID: deadline reminders use a
    deterministic id (``{opportunity_id}-deadline-{days}``) so re-scheduling
    is idempotent by primary-key collision, matching the Dart demo
    repository's id-equality dedup. Event/admin-created records use a random
    hex id instead.
    """

    __tablename__ = "scholarsphere_notifications"
    __table_args__ = (
        Index("ix_notifications_user_status_scheduled", "user_id", "status", "scheduled_for"),
        Index("ix_notifications_status_scheduled", "status", "scheduled_for"),
    )

    id: Mapped[str] = mapped_column(String(255), primary_key=True)
    user_id: Mapped[str] = mapped_column(String(255), nullable=False, index=True)
    type: Mapped[NotificationEventType] = mapped_column(
        Enum(NotificationEventType, native_enum=False), nullable=False
    )
    title: Mapped[str] = mapped_column(String(500), nullable=False)
    message: Mapped[str] = mapped_column(Text, nullable=False)
    channels: Mapped[list[str]] = mapped_column(JSONType, default=list, nullable=False)
    scheduled_for: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    opportunity_id: Mapped[UUID | None] = mapped_column(
        ForeignKey("external_opportunities.id", ondelete="RESTRICT")
    )
    read_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    template_id: Mapped[str | None] = mapped_column(
        ForeignKey("notification_templates.id", ondelete="SET NULL")
    )
    related_entity_type: Mapped[str | None] = mapped_column(String(64))
    related_entity_id: Mapped[str | None] = mapped_column(String(255))
    sent_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    delivered_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    status: Mapped[NotificationDeliveryStatus] = mapped_column(
        Enum(NotificationDeliveryStatus, native_enum=False),
        default=NotificationDeliveryStatus.scheduled,
        nullable=False,
    )
    retry_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    failure_reason: Mapped[str | None] = mapped_column(Text)
    timezone: Mapped[str] = mapped_column(String(64), default="UTC", nullable=False)
    group_key: Mapped[str | None] = mapped_column(String(255))
